File: 백준/레벨Gold IV/2179. 비슷한 단어/비슷한 단어.ts
// @ts-ignore
import fs = require('fs');

const solution = () => {
  const read = fs.readFileSync(0, 'utf-8').split('\n');

  const N = parseInt(read[0].trim());

  // 사전 순 정렬
  const words = read
    .slice(1)
    .map((word: string, idx: number) => [word.trim(), idx + 1] as [string, number])
    .sort((a: [string, number], b: [string, number]) => a[0].localeCompare(b[0]));

  // 인접 글자 간 LCP 처리
  let maxLen = 0;
  const result = new Map<number, string>();

  for (let i = 0; i < words.length - 1; i++) {
    const [pre, preIdx] = words[i];
    const [nxt, nxtIdx] = words[i + 1];
    const size = Math.min(pre.length, nxt.length);

    // LCP 길이 확인
    let curLen = 0;
    for (let j = 0; j < size; j++) {
      if (pre[j] === nxt[j]) curLen++;
      else break;
    }

    // LCP 갱신
    if (maxLen < curLen) {
      const lcp = pre.slice(0, curLen);
      maxLen = curLen;
      result.clear();
      result.set(preIdx, lcp);
      result.set(nxtIdx, lcp);
    } else if (maxLen === curLen) {
      const lcp = pre.slice(0, curLen);
      result.set(preIdx, lcp);
      result.set(nxtIdx, lcp);
    }
  }

  // 여러 개일 때에는 입력되는 순서대로
  const idxList = Array.from(result.entries()).sort((a, b) => a[0] - b[0]);
  const [idx1, common] = idxList[0];

  let idx2 = -1;
  for (let i = 1; i < idxList.length; i++) {
    if (idxList[i][1] === common) {
      idx2 = idxList[i][0];
      break;
    }
  }

  console.log(read[idx1]);
  console.log(read[idx2]);
};

solution();

File: 백준/레벨Gold IV/2179. 비슷한 단어/비슷한 단어.java
/*
    골드4 - 2179번: 비슷한 단어  https://www.acmicpc.net/problem/2179
 */

import java.io.*;
import java.util.*;

class Word {
    String text;
    int index;
    
    Word(String text, int index) {
        this.text = text;
        this.index = index;
    }
}

class LCP {
    String text;
    int index;

    LCP(String text, int index) {
        this.text = text;
        this.index = index;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || this.getClass() != o.getClass()) return false;
        LCP lcp = (LCP) o;
        return Objects.equals(this.text, lcp.text) &&
               this.index == lcp.index;
    }

    @Override
    public int hashCode() {
        return Objects.hash(text, index);
    }
}

class Solution {
    public void solution() throws IOException  {
        String[] read = new String(System.in.readAllBytes()).split("\n");
        int N = Integer.parseInt(read[0].trim());

        // 사전순 정렬된 wordList를 생성
        List<Word> wordList = new ArrayList<>(N);
        for (int i = 1; i < read.length; i++) {
            wordList.add(new Word(read[i].trim(), i));
        }
        wordList.sort(Comparator.comparing(w -> w.text));

        // 인접 단어 사이의 LCP 중 가장 긴 것을 추출
        int maxLen = 0;
        Set<LCP> lcpSet = new HashSet<>();

        for (int i = 0; i < wordList.size() - 1; i++) {
            String pre = wordList.get(i).text;
            int pIdx = wordList.get(i).index;

            String nxt = wordList.get(i + 1).text;
            int nIdx = wordList.get(i + 1).index;

            // LCP의 길이 확인
            int curLen = 0;
            int size = Math.min(pre.length(), nxt.length());
            for (int j = 0; j < size; j++) {
                if (pre.charAt(j) == nxt.charAt(j)) {
                    curLen++;
                } else {
                    break;
                }
            }

            // 최대 길이가 갱신되면 SET 교체
            if (maxLen < curLen) {
                maxLen = curLen;
                String text = pre.substring(0, curLen);
                lcpSet.clear();
                lcpSet.add(new LCP(text, pIdx));
                lcpSet.add(new LCP(text, nIdx));
            }
            // 최대 길이가 동일하면 SET에 추가
            else if (maxLen == curLen) {
                String text = pre.substring(0, curLen);
                lcpSet.add(new LCP(text, pIdx));
                lcpSet.add(new LCP(text, nIdx));
            }
        }

        // Set -> List, 입력 순 정렬
        List<LCP> lcpList = new ArrayList<>(lcpSet);
        lcpList.sort(Comparator.comparingInt(l -> l.index));
        
        String lcp = lcpList.get(0).text;
        int idx1 = lcpList.get(0).index;

        int idx2 = -1;
        for (int i = 1; i < lcpList.size(); i++) {
            String cur = lcpList.get(i).text;
            if (cur.equals(lcp)) {
                idx2 = lcpList.get(i).index;
                break;
            }
        }

        System.out.println(read[idx1]);
        System.out.println(read[idx2]);
    }
}

public class Main {
    public static void main(String[] args) throws IOException  {
        new Solution().solution();
    }
}